### ISSUE TITLE
Slack notify task not running

### ISSUE THREAD

**Issue by bocanhcam** [deployphp/deployer#2100]
Everything is good except `slack:notify` and `slack:notify:success` seem like not working at all.

I checked the code inside `recipe/slack.php` and I think this line is not working

`Httpie::post(get('slack_webhook'))->body(['attachments' => [$attachment]])->send();
`

Can anybody help me solved this?

**Environment**
* Deployer version: v6.8.0
* PHP version: 7.2.30
* Deployment target(s) OS: Ubuntu 18.04

**Content of `deploy.php`**
```php
<?php
namespace Deployer;

require 'recipe/laravel.php';
require 'recipe/rsync.php';
require 'recipe/slack.php';

set('allow_anonymous_stats', false);

set('ssh_type', 'native');
set('ssh_multiplexing', true);

set('writable_use_sudo', true);
set('writable_mode', 'chmod');
set('writable_chmod_mode', '0777');
set('writable_chmod_recursive', true);
set('http_user', 'deployer:nginx');

// Project name
set('application', 'icc_project');

// Project repository
set('repository', 'git@my_git_lab_project.git');
set('ssh_multiplexing', true);

// [Optional] Allocate tty for git clone. Default value is false.
set('git_tty', true);

set('branch', 'develop');
set('ssh_type', 'native');
set('writable_use_sudo', true);


// Shared files/dirs between deploys
add('shared_files', []);
add('shared_dirs', []);

// Writable dirs by web server
add('writable_dirs', []);


// slack notify
set('slack_webhook', 'https://hooks.slack.com/services/my_key_xxx');
// Deploy message
set('slack_text', '_{{user}}_ deploying `{{branch}}` to *{{target}}*');
set('slack_success_text', 'Deploy to *{{target}}* successful');
set('slack_failure_text', 'Deploy to *{{target}}* failed');

// Color of attachment
set('slack_color', '#4d91f7');
set('slack_success_color', '#00c100');
set('slack_failure_color', '#ff0909');

// Hosts

host('amazon-server')
    ->port(22)
    ->set('deploy_path', '/var/www/html/icc_project')
    ->configFile('~/.ssh/config')
    ->forwardAgent(true)
    ->multiplexing(true)
    ->addSshOption('UserKnownHostsFile', '/dev/null')
    ->addSshOption('StrictHostKeyChecking', 'no');


set('composer_options', 'install --verbose --prefer-dist --no-progress --no-interaction --optimize-autoloader --no-suggest');


task('test', function () {
    writeln('Hello world');
});

// supervisor
task('supervisor:restart', function () {
    run('sudo service supervisord restart');
})->onRoles('root')->desc('Restart supervisord restart');

task('deploy', [
    'deploy:prepare',
    //'deploy:lock',
    'deploy:release',
    'deploy:update_code',
    'deploy:shared',
    'deploy:writable',
    'deploy:vendors',
    'deploy:clear_paths',
    'deploy:symlink',
    'artisan:migrate',
    'artisan:cache:clear',
    'artisan:view:clear',
    'supervisor:restart',
    //'deploy:unlock',
    'cleanup',
    'success'
]);

after('success', 'test');

// slack
before('deploy', 'slack:notify');
after('success', 'slack:notify:success');
after('deploy:failed', 'slack:notify:failure');

// [Optional] if deploy fails automatically unlock.
after('deploy:failed', 'deploy:unlock');
```

**Output log with enabled option for verbose output `-vvv`**
```sh
...
➤ Executing task artisan:view:clear
[amazon-server] > /usr/bin/php /var/www/html/icc_project/releases/10/artisan view:clear
[amazon-server] < Compiled views cleared!
• done on [amazon-server]
✔ Ok [539ms]
➤ Executing task cleanup
[amazon-server] >  rm -rf /var/www/html/icc_project/releases/5
[amazon-server] > cd /var/www/html/icc_project && if [ -e release ]; then  rm release; fi
[amazon-server] > cd /var/www/html/icc_project && if [ -h release ]; then  rm release; fi
• done on [amazon-server]
✔ Ok [2s 53ms]
Successfully deployed!
➤ Executing task test
Hello world
• done on [amazon-server]
✔ Ok [0ms]
• done on [amazon-server]
```

**Comment by roberto910907** [comment]
@bocanhcam the same is happening to me!
TL:DR there's a timeout issue in the slack recipe. As a workaround you can create your own recipe by copying the original one and increasing the timeout value.

```
desc('My own task to send notifications');
task('slack:notify:failure', function () {
    if (!get('slack_webhook', false)) {
        return;
    }

    $attachment = [
        'title' => get('slack_title'),
        'text' => get('slack_failure_text'),
        'color' => get('slack_failure_color'),
        'mrkdwn_in' => ['text'],
    ];

    Httpie::post(get('slack_webhook'))->setopt(CURLOPT_TIMEOUT, 50)->setopt(CURLOPT_CONNECTTIMEOUT, 50)->body(['attachments' => [$attachment]])->send();
})
    ->once()
    ->shallow()
    ->setPrivate();
```

Please notice that I've changed this line:
From:
```
Httpie::post(get('slack_webhook'))->body(['attachments' => [$attachment]])->send();
```

To:
```
Httpie::post(get('slack_webhook'))->setopt(CURLOPT_TIMEOUT, 50)->setopt(CURLOPT_CONNECTTIMEOUT, 50)->body(['attachments' => [$attachment]])->send();
```
**That's it!**

-----------------------------------------------------------------------------------------------------------------------------------------
**Explanation:** 

I found the root cause for this issue is the Deployer [Httpie](https://github.com/deployphp/deployer/blob/master/src/Utility/Httpie.php) class where slack recipe rely on. It turns out that this class has a fix timeout value of 5 seconds and probably the slack request is taking more than that to be executed.

Just to verify this I've added these lines [here](https://github.com/deployphp/deployer/blob/master/src/Utility/Httpie.php#L101-L102):
```
$result = curl_exec($ch);
var_dump(curl_error($ch));
```
And the result:
**string(43) "Resolving timed out after 5513 milliseconds"**

Fortunately they have a method that allows us to overwrite any CURL constant configuration, and, my proposed solution is based on that.

Unfortunately, AFAIK, the slack recipe does not have a way to increase those value directly.

Let me know if this solutions is working for you! 
Cheers.

**Comment by asafov** [comment]
Resolving timed out means, that you have troubles with DNS resolve.
Which DNS server use using now? 5 seconds timeout for send 1 simple request isn't a small value.
Seems that you have network issues.

**Comment by roberto910907** [comment]
@asafov This is actually failing on CircleCI building, unfortunately I'm unable to help them to solve their network issues :)

**Comment by bocanhcam** [comment]
Thank you @roberto910907
it working for me too

**Comment by bocanhcam** [comment]
@asafov 
This is what I used.

NS               ns1.matbao.com.
                    ns2.matbao.com.

Btw, My project is on AWS EC2. Is it better to use route53?

**Comment by asafov** [comment]
Tru to use 8.8.8.8 as DNS resolver. Maybe it will help.

**Comment by antonmedv** [comment]
I am cleaning up on issues. Fill free to reopen if needed. 
Also, we have a new section available: [discussions](https://github.com/deployphp/deployer/discussions).
